security/fma-multi-threaded-64-bit-cannon.md

```diff
@@ -0,0 +1,164 @@
+- **Risk Assessment:** High severity, low likelihood.
+- **Mitigations:** Manual review of the op-program and a quick review of Go runtime internals. The op-program uses 3 threads, and only one of those threads is used by the mutator main function. This makes livelocks very unlikely. This [issue](https://github.com/ethereum-optimism/optimism/issues/11979) looks into the livelock problem with possible solutions. The proposed solutions are deferred for future work as the risk of a livelock is considered too low to be addressed immediately.
+- **Detection:** This would manifest as an execution that runs forever. Eventually, but well before the dispute period ends, op-dispute-mon will indicate that a game is forecasted to resolve incorrectly.
+- **Recovery Path(s)**: See [Fault Proof Recovery](https://www.notion.so/oplabs/RB-000-Fault-Proofs-Recovery-Runbook-8dad0f1e6d4644c281b0e946c89f345f).
+
+### Execution traces too long for the fault proof
+
+- **Description:** It's possible that introducing multi-threading/gc greatly increases the execution time of the op-program.
+- **Risk Assessment:** Medium severity, low likelihood.
+- **Mitigations:** Based on [vm-runner executions of 64-bit Cannon and 32-bit single-threaded Cannon](https://optimistic.grafana.net/goto/IlD3Tu4Hg?orgId=1), the 64-bit VM executes the op-program much faster than the 32-bit VM. As such, we do not expect worse performance over the existing VM, which was already found to be adequate. However, we can always use CPUs with better single-core performance to mitigate.
+- **Detection:** op-dispute-mon provides an early forecast (see [this alerting rule](https://github.com/ethereum-optimism/k8s/blob/6416658dd9602f83c6be093c5c57af45877258a3/grafana-cloud/rules/dispute-mon.yaml#L154)) and triggers an alert if the op-challenger stops interacting with a game.
+- **Recovery Path(s)**: This can be mitigated by migrating the op-challenger to a more powerful CPU. [Fault Proof Recovery](https://www.notion.so/oplabs/RB-000-Fault-Proofs-Recovery-Runbook-8dad0f1e6d4644c281b0e946c89f345f) provides guidance on when it'll be appropriate to do so.
+
+### Invalid `DisputeGameFactory.setImplementation` execution
+
+The `DisputeGameFactory.setImplementation` will be used to upgrade both the `CANNON` and `PERMISSIONED_CANNON` game types. Note that this means this upgrade will not require changing the `respectedGameType` in the `OptimismPortal` as `CANNON` games will continue to be used to finalize outputs, albeit with a different VM.
+As such, there will be a brief moment where there are two sets of `CANNON` games that using singlethreaded and multithreaded VMs.
+
+- Description: This occurs when either the call to the DisputeGameFactory could not be made due to grossly unfavorable base fees on L1, an invalidly approved safe nonce, or a successful execution to a misconfigured dispute game implementation.
```

Comment:
I don't understand this failure mode. Do you mean that there will be unfinalized games of both types for a brief moment of up to a few days (until the very last ST-Cannon game resolves)? The "successful execution to a misconfigured dispute game implementation" aspect makes sense, but the rest of it I don't understand. Would the transaction just not be executed if we e.g. signed the wrong nonce?

Reply:
I don't understand either. @mbaxter can you clarify? To me this sounds like two things:
1. there's a short period when single-threaded games might still be in progress (not a problem I think?)
2. if the upgrade transaction fails that period will last until we re-do the transaction? (not really a failure mode that needs to be documented in here IMO)

Reply:
I think this section (much of which comes from [this template](https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-generic-hardfork.md#invalid-disputegamefactorysetimplementation-execution)) is just describing what could go wrong with `setImplementation`:
- We misconfigure it and set the wrong implementation (unlikely due to review)
- The transaction fails to execute due to gas fees, or a bad nonce

I think the bit about multiple types of games is just informational - no real issues around this.  And yes, I think the description from @mds1 is right:
> there will be unfinalized games of both types for a brief moment of up to a few days (until the very last ST-Cannon game resolves)

Reply:
@mds1 Does this make sense? I think it's not a very useful section. We could remove it but I vote to just resolve this comment and leave it.

Reply:
I don't know if we can remove the full section but after seeing the the template linked by @mbaxter I do understand that if we still have an `challenger` that run a `op-program` that run the previous version we can still challenge as expected a malicious claim.

I don't know if this possible to have a process during hardfork to make sure we also keep the previous `op-program` ready to run in case we miss the upgrade as a fallback mechanism?

Reply:
Right - for this FMA we aren't dealing with a hardfork, so the op-challenger will continue to operate normally if we were to somehow miss the `setImplementation` call. 

During a hardfork, we have to upgrade ahead of the hardfork activation, otherwise the fault proof system will be running the wrong consensus rules. So having the old version running doesn't help in that case.

Reply:
Oh ok, this is more clear for me for now!
I think we can close and resolve this. 
Also, we can said in the mitigation that this is different that hardfork this will still be able to operate normally if needed?

Reply:
Thanks for the clarification @mbaxter. Since a few of us were confused by this failure mode we should update the FMA text to clarify and answer @Ethnical's question, then we can resolve this thread

Reply:
I think we should just remove it. It's a generic failure mode, documented n the generic hardfork FMA, we're not going to do anything new to mitigate this for MT Cannon (are we?). Feel free to propose an edit, but if it's on me to do that it's going to keep hanging around and I'd love to get this FMA closed.

Reply:
Ok @pauldowman, I have commit the suggestion here just to make sure this is clear enough for everyone in the future ->
https://github.com/ethereum-optimism/design-docs/pull/123/files#r1954545791
if you can commit the suggestion and we can resolve this comment and we can merge today :)